From 7b3dbea2bcbb6501465ee46ef16e188342ab24ca Mon Sep 17 00:00:00 2001
From: Sunil Goutham <sgoutham@cavium.com>
Date: Tue, 9 Oct 2018 00:42:32 +0300
Subject: [PATCH 0259/1051] ptp: Expose PHC clock to network driver

This patches exposes PHC clock for network driver to use.

This is not how it should be done, this will be reworked pretty soon
for a upstream acceptable version. Due to urgent customer requirment on
81xx it has been done this way.

Signed-off-by: Yury Norov <ynorov@caviumnetworks.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/ptp/cavium_ptp.c          |  6 +++++-
 drivers/ptp/cavium_ptp.h          | 11 +++++++++++
 drivers/ptp/cavium_thunderx_ptp.c | 28 ++++++++++++++++++++--------
 3 files changed, 36 insertions(+), 9 deletions(-)

diff --git a/drivers/ptp/cavium_ptp.c b/drivers/ptp/cavium_ptp.c
index fce7507754de..045d9a2c3fd5 100644
--- a/drivers/ptp/cavium_ptp.c
+++ b/drivers/ptp/cavium_ptp.c
@@ -108,7 +108,11 @@ static int cavium_ptp_adjtime(struct ptp_clock_info *ptp_info, s64 delta)
 
 	/* notify child module about the time adjust */
 	if (cavium_ptp_info->adjtime_clbck)
-		cavium_ptp_info->adjtime_clbck(cavium_ptp_info, delta);
+		cavium_ptp_info->adjtime_clbck(cavium_ptp_info,
+					       cavium_ptp_clock->ptp_adjust);
+
+	/* Sync, for network driver to get latest value */
+	smp_mb();
 
 	return 0;
 }
diff --git a/drivers/ptp/cavium_ptp.h b/drivers/ptp/cavium_ptp.h
index 97889e4e65e7..f1728ba85789 100644
--- a/drivers/ptp/cavium_ptp.h
+++ b/drivers/ptp/cavium_ptp.h
@@ -46,6 +46,17 @@ struct cavium_ptp_clock {
 	struct cavium_ptp_clock_info *cavium_ptp_info;
 };
 
+struct thunder_ptp_clock {
+	void __iomem *reg_base;
+	struct pci_dev *pdev;
+	struct cavium_ptp_clock *cavium_ptp_clock;
+	struct cavium_ptp_clock_info cavium_ptp_info;
+	s64 ptp_adjust;
+};
+
+extern struct thunder_ptp_clock *thunder_ptp_clock;
+s64 thunder_get_adjtime(void);
+
 extern struct cavium_ptp_clock *cavium_ptp_register(
 	struct cavium_ptp_clock_info *info, struct device *dev);
 extern void cavium_ptp_remove(struct cavium_ptp_clock *cavium_ptp_clock);
diff --git a/drivers/ptp/cavium_thunderx_ptp.c b/drivers/ptp/cavium_thunderx_ptp.c
index 81076c63e2f1..2772e33821c3 100644
--- a/drivers/ptp/cavium_thunderx_ptp.c
+++ b/drivers/ptp/cavium_thunderx_ptp.c
@@ -33,12 +33,7 @@
 /* PCI device IDs */
 #define PCI_DEVICE_ID_THUNDER_PTP	0xA00C
 
-struct thunder_ptp_clock {
-	void __iomem *reg_base;
-	struct pci_dev *pdev;
-	struct cavium_ptp_clock *cavium_ptp_clock;
-	struct cavium_ptp_clock_info cavium_ptp_info;
-};
+struct thunder_ptp_clock *thunder_ptp_clock;
 
 /*
  * Register access functions
@@ -72,13 +67,30 @@ static void thunder_ptp_reg_write(struct cavium_ptp_clock_info *info,
 	writeq_relaxed(val, addr);
 }
 
+static void thunder_ptp_adjtime(struct cavium_ptp_clock_info *info,
+				   s64 delta)
+{
+	struct thunder_ptp_clock *thunder_ptp_clock =
+		container_of(info, struct thunder_ptp_clock, cavium_ptp_info);
+
+	thunder_ptp_clock->ptp_adjust = delta;
+}
+
+s64 thunder_get_adjtime(void)
+{
+	if (!thunder_ptp_clock)
+		return 0;
+
+	return thunder_ptp_clock->ptp_adjust;
+}
+EXPORT_SYMBOL(thunder_get_adjtime);
+
 /* module operations */
 
 static int thunder_ptp_probe(struct pci_dev *pdev,
 				 const struct pci_device_id *ent)
 {
 	int err;
-	struct thunder_ptp_clock *thunder_ptp_clock = NULL;
 	struct device *dev = &pdev->dev;
 
 	thunder_ptp_clock = devm_kzalloc(dev, sizeof(*thunder_ptp_clock),
@@ -119,7 +131,7 @@ static int thunder_ptp_probe(struct pci_dev *pdev,
 		.name = "ThunderX PTP",
 		.reg_read = thunder_ptp_reg_read,
 		.reg_write = thunder_ptp_reg_write,
-		.adjtime_clbck = NULL,
+		.adjtime_clbck = thunder_ptp_adjtime,
 	};
 	thunder_ptp_clock->cavium_ptp_clock = cavium_ptp_register(
 		&thunder_ptp_clock->cavium_ptp_info, dev);
-- 
2.17.1

